### PR TITLE
Added SystemParticipantWithHeatResult class and modified ChpResult accordingly

### PR DESCRIPTION
**Pull request by ahamshubham**
Resolves ie3-institute/PowerSystemDataModel#236. Changes done -
1) Created a class `SystemParticipantWithHeatResult` to include the additional thermal power variable `qDot`.
2) Extended `ChpResult.java` to inherit this class.
3) Modified `SystemParticipantResultFactory.java` to add the additional thermal power variable.

Several test cases are failing now. I am trying to resolve the issues.

### CODE REVIEW

**Comment by ahamshubham** [comment]
I have modified the test cases according to the new code. But, there are some codacy static code analysis issues which I have fixed but they are being shown. And there are several test cases which are failing.

**Comment by ahamshubham** [comment]
Regarding the test cases which are failing, I tried running `./gradlew test` on the `dev` branch, the 5 test cases which fail on my branch `sb/#236-modifyingChpResult` are also failing on the `dev` branch. The Codacy Static Code Analysis test is successful now. I made a mistake while understanding the error messages before.

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

**Comment by ahamshubham** [comment]
Need to add more test cases to test `SystemParticipantWithHeatResult` class probably to increase the code coverage.

**Comment by ahamshubham** [comment]
!test

**Review by ckittl — CHANGES_REQUESTED** [review]
Thank you for your work on this. Please find some remarks below.

Additionally, please let `HpResult` also be a child of `SystemParticipantWithHeatResult` and remove the `qDot`-attribute there. Finally, please adapt the uml documentation for result classes (OutputDataModelConcept.puml) as well as the corresponding section in Sphinx-Documentation (docs/readthedocs/models/result/participant/chp.rst)

File: src/main/java/edu/ie3/datamodel/models/result/system/ChpResult.java [hunk 1]
```diff
@@ -11,7 +11,9 @@
 import tech.units.indriya.ComparableQuantity;
 
 /** Represents calculation results of a {@link edu.ie3.datamodel.models.input.system.ChpInput} */
-public class ChpResult extends SystemParticipantResult {
+public class ChpResult extends SystemParticipantWithHeatResult {
+
+  private final ComparableQuantity<Power> qDot;
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/ChpResult.java:16]: Sure, I will remove it.

File: src/main/java/edu/ie3/datamodel/models/result/system/ChpResult.java [hunk 2]
```diff
@@ -61,6 +69,8 @@ public String toString() {
         + getP()
         + ", q="
         + getQ()
+        + ", qDot="
+        + qDot
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/ChpResult.java:73]: Sure.

File: src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+package edu.ie3.datamodel.models.result.system;
+
+import java.time.ZonedDateTime;
+import java.util.UUID;
+import javax.measure.quantity.Power;
+import tech.units.indriya.ComparableQuantity;
+
+/** Abstract class that holds values common to result entities having heat result */
+public class SystemParticipantWithHeatResult extends SystemParticipantResult {
+  /** @param qDot thermal power output normally provided in MW */
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java:15]: Sure, I will remove it and modify the JavaDoc.

File: src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  private ComparableQuantity<Power> qDot;
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java:16]: I will make the change.

File: src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+*/
+package edu.ie3.datamodel.models.result.system;
+
+import java.time.ZonedDateTime;
+import java.util.UUID;
+import javax.measure.quantity.Power;
+import tech.units.indriya.ComparableQuantity;
+
+/** Abstract class that holds values common to result entities having heat result */
+public class SystemParticipantWithHeatResult extends SystemParticipantResult {
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java:14]: Sure, I will make the change.

File: src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        + getInputModel()
+        + ", p="
+        + getP()
+        + ", q="
+        + getQ()
+        + ", qDot="
+        + getQDot()
+        + '}';
+  }
+}
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/models/result/system/SystemParticipantWithHeatResult.java:82]: Sure, I will add equals and hasCode methods.

File: src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java [hunk 7]
```diff
@@ -55,7 +55,8 @@ public SystemParticipantResultFactory() {
       optionalFields = expandSet(minConstructorParams, ENTITY_UUID);
     }
 
-    if (data.getTargetClass().equals(HpResult.class)) {
+    if (data.getTargetClass().equals(HpResult.class)
+        || data.getTargetClass().equals(ChpResult.class)) {
       minConstructorParams = newSet(TIME, INPUT_MODEL, POWER, REACTIVE_POWER, Q_DOT);
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java:60]: Sure, I will make the change.

File: src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java [hunk 8]
```diff
@@ -96,9 +97,10 @@ protected SystemParticipantResult buildModel(SimpleEntityData data) {
           .map(uuid -> new EvcsResult(uuid, zdtTime, inputModelUuid, p, q))
           .orElseGet(() -> new EvcsResult(zdtTime, inputModelUuid, p, q));
     } else if (entityClass.equals(ChpResult.class)) {
+      ComparableQuantity<Power> qDot = data.getQuantity(Q_DOT, StandardUnits.Q_DOT_RESULT);
       return uuidOpt
-          .map(uuid -> new ChpResult(uuid, zdtTime, inputModelUuid, p, q))
-          .orElseGet(() -> new ChpResult(zdtTime, inputModelUuid, p, q));
+          .map(uuid -> new ChpResult(uuid, zdtTime, inputModelUuid, p, q, qDot))
+          .orElseGet(() -> new ChpResult(zdtTime, inputModelUuid, p, q, qDot));
     } else if (entityClass.equals(WecResult.class)) {
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java:104]: Sure, I will make the change.

File: src/test/groovy/edu/ie3/datamodel/models/result/system/ChpResultTest.groovy [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/*
+ * © 2021. TU Dortmund University,
+ * Institute of Energy Systems, Energy Efficiency and Energy Economics,
+ * Research group Distribution grid planning and operation
+ */
+package edu.ie3.datamodel.models.result.system
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/datamodel/models/result/system/ChpResultTest.groovy:6]: Sure, I will do that.

File: src/test/groovy/edu/ie3/datamodel/io/sink/InfluxDbSinkIT.groovy [hunk 10]
```diff
@@ -27,6 +27,7 @@ import spock.lang.Shared
 import spock.lang.Specification
 import tech.units.indriya.quantity.Quantities
 
+import java.nio.file.StandardCopyOption
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/datamodel/io/sink/InfluxDbSinkIT.groovy:30]: I will remove it.

File: src/test/groovy/edu/ie3/test/common/SystemParticipantTestData.groovy [hunk 11]
```diff
@@ -324,4 +324,12 @@ class SystemParticipantTestData {
...
 	]
 
+	//ChpResult & SystemParticipantWithHeatResult
+	private static final UUID uuid = UUID.fromString("22bea5fc-2cb2-4c61-beb9-b476e0107f52")
+	private static final ZonedDateTime time = java.time.ZonedDateTime.parse("2020-01-30T17:26:44Z[UTC]")
+	private static final UUID inputModel = UUID.fromString("22bea5fc-2cb2-4c61-beb9-b476e0107f52")
+	private static final Quantity<Power> p = Quantities.getQuantity(10, StandardUnits.ACTIVE_POWER_IN)
+	private static final Quantity<Power> q = Quantities.getQuantity(10, StandardUnits.REACTIVE_POWER_IN)
+	private static final Quantity<Power> qDot = Quantities.getQuantity(1, StandardUnits.Q_DOT_RESULT)
+
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/test/groovy/edu/ie3/test/common/SystemParticipantTestData.groovy:334]: I will remove it.

File: src/main/java/edu/ie3/datamodel/io/processor/Processor.java [hunk 12]
```diff
@@ -137,6 +139,12 @@ public int compare(String a, String b) {
                   fieldName = V_RATED;
                   resFieldNameToMethod.put(VOLT_LVL, pd.getReadMethod());
                 }
+
+                // Rename the fieldName for thermal power
+                if (fieldName.equalsIgnoreCase(Q_DOT)) {
+                  fieldName = Q_DOT;
+                }
```
[the remark this answers is not in the document]
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/processor/Processor.java:146]: Sure, I will make the change.

**Comment by ahamshubham** [comment]
Actually, I have made the suggested changes (except for the changes to the uml and Sphinx documentation) to the pull request 269 but the problem is that there is Quality Gate failure. I believe the problem is that the added functions in the class SystemParticipantWithHeatResult.java have not been tested.

File: docs/uml/main/OutputDatamodelConcept.puml [hunk 13]
```diff
@@ -82,10 +82,14 @@ package models {
             class PvResult
             SystemParticipantResult  <|-- PvResult
 
-            class ChpResult
+            class ChpResult{
+                + qDot: ComparableQuantity<Power>
+            }
             SystemParticipantResult <|-- ChpResult
```
Comment (ckittl) [inline docs/uml/main/OutputDatamodelConcept.puml:88]: The UML documentation is not correct. `ChpResult` and `HpResult` inherit from `SystemParticipantWithHeatResult`. That one is missing completely. Additionally, `qDot` then is part of `SystemParticipantWithHeatResult`.

File: src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java [hunk 14]
```diff
@@ -57,7 +57,7 @@ public SystemParticipantResultFactory() {
 
     if (data.getTargetClass().equals(HpResult.class)
         || data.getTargetClass().equals(ChpResult.class)) {
```
Comment (ckittl) [inline src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java]: Additional remark: You can simplify this for future cases, where classes extend `SystemParticipantWithHeatResult`:

```suggestion
    if (SystemParticipantWithHeatResult.class.isAssignableFrom(data.getTargetClass()) {
```

**Review by ckittl — CHANGES_REQUESTED** [review]

File: src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java (hunk 14 above)
Comment (ahamshubham) [reply src/main/java/edu/ie3/datamodel/io/factory/result/SystemParticipantResultFactory.java]: Sure, I will make the change.

File: docs/uml/main/OutputDatamodelConcept.puml (hunk 13 above)
Comment (ahamshubham) [reply docs/uml/main/OutputDatamodelConcept.puml:88]: Sure, I will make the changes.

**Comment by ahamshubham** [comment]
!test

**Comment by ahamshubham** [comment]
I have added the requested changes to the puml file and the `SystemParticipantResultFactory` class. I also modified `AUTHORS` and `CHANGELOG.md`. However, when I was making the first two changes, I realized that in `OutputDataModelConcept.puml`, `StorageResult` and `EvResult` are defined in the same way I did previously for `ChpResult` and `HpResult`. So, I believe we need to change that. Also, a similar change to what you suggested for the if condition containing `ChpResult` and `HpResult` in `SystemParticipantResultFactory` class can be applied for the if condition containing `StorageResult` and `EvResult`.

Right now, several test cases are failing here but they all seem to pass on my local machine.

**Comment by ahamshubham** [comment]
!test

**Comment by ckittl** [comment]
> I have added the requested changes to the puml file and the `SystemParticipantResultFactory` class. I also modified `AUTHORS` and `CHANGELOG.md`. However, when I was making the first two changes, I realized that in `OutputDataModelConcept.puml`, `StorageResult` and `EvResult` are defined in the same way I did previously for `ChpResult` and `HpResult`. So, I believe we need to change that. Also, a similar change to what you suggested for the if condition containing `ChpResult` and `HpResult` in `SystemParticipantResultFactory` class can be applied for the if condition containing `StorageResult` and `EvResult`.
> 
> Right now, several test cases are failing here but they all seem to pass on my local machine.

Thanks for that catch! I would suggest to only stick to one issue here (this is adding the heat part). This makes it easier to keep track of PRs.

**Comment by ckittl** [comment]
!test

**Review by ckittl — APPROVED** [review]
Please wait with merging, until the tests are passing.

**Review by ckittl — APPROVED** [review]

**Pull request merged**